### ISSUE TITLE
Populate History Automatically

### ISSUE THREAD

**Issue by emptyrivers** [WeakAuras/WeakAuras-Companion#390]
In WeakAuras version 2.14 and later, there exists a history collection. This data will be used for lots of neat things, but its usefulness is limited by the availability of that data.

For users who have the companion app installed, we can accelerate the accumulation of this data by pulling previous versions from wago, and retroactively populating history.

The current available endpoint is as follows:
```lua
WeakAuras.SetHistory(uid, snapshot, source, addon)
```
With `uid` being the unique identifier for the relevant aura whose history is being populated, `snapshot` being the previous version of said aura, `source` the source of the history (used for display purposes), and `addon` indicating the addon from which this history originates.

What do you guys think? Is this endpoint usable for the companion app to backfill history, or should we extend it or introduce something new?

**Comment by mrbuds** [comment]
Would that be a one time thing ?

**Comment by mrbuds** [comment]
What would be the requirement on the data of an aura that would trigger companion's update process to pull data from wago for back-filling history ?

**Comment by mrbuds** [comment]
Or better, is it possible to add a "flag" like `data.backfillHistory = true` that would force companion to get the data and write it in another table to be parsed by the addon ?
I'm asking this because parsing the SV from js lua lib is not the easiest task.

**Comment by emptyrivers** [comment]
> Would that be a one time thing ?

Effectively, yes. But I don't see any way to enforce that which would not also add moving parts that easily break.

> What would be the requirement on the data of an aura that would trigger companion's update process to pull data from wago for back-filling history ?
```js
foreach (data in WeakAurasSaved.displays) {
  if (data.uid !== null && WeakAurasSaved.history[data.uid] === null && data.url !== null) {
    collectHistoricalData(data.url);
  }
}
```

> Or better, is it possible to add a "flag" like `data.backfillHistory = true` that would force companion to get the data and write it in another table to be parsed by the addon ?
> I'm asking this because parsing the SV from js lua lib is not the easiest task.

I do not wish to add more companion-specific flags to WeakAuras data, no.

**Comment by InfusOnWoW** [comment]
I'm not sure, that sounds like a lot of data to add to WeakAuras.lua for a not so significant feature.

**Comment by mrbuds** [comment]
@emptyrivers i made an update so companion get all missing encoded string on it's db from wago, and instead of only writing data for auras with update in WeakAurasCompanion addon's data.lua it now write data for all auras with an encoded string (no need to make any change on WeakAuras side to handle it)

So now i guess i need to make Companion addon backfill history, if we still agree on doing it, can you check that this is right ?
```Lua
for id, data in pairs(WeakAurasSaved.displays) do
   if data.uid and not WeakAurasSaved.history[data.uid] then
      local slug = WeakAurasCompanion.uids[data.uid]
      if slug then
         local wagoData = WeakAurasCompanion.slugs[slug]
         if wagoData and wagoData.encoded then
            WeakAuras.SetHistory(data.uid, wagoData.encoded, "Wago", "WeakAurasCompanion")
         end
      end
   end
end
```

**Comment by mrbuds** [comment]
An example of Companion data with this change
[WeakAurasCompanion.zip](https://github.com/WeakAuras/WeakAuras-Companion/files/3718790/WeakAurasCompanion.zip)

**Comment by mrbuds** [comment]
This is implemented in version 2.0.0-beta1
